mkosi.default.d/10-systemd.conf

```diff
@@ -6,6 +6,8 @@
 Format=gpt_btrfs
 Bootable=yes
 HostonlyInitrd=yes
+KernelCommandLine=systemd.setenv=ASAN_OPTIONS=strict_string_checks=1:detect_stack_use_after_return=1:check_initialization_order=1:strict_init_order=1:disable_coredump=0:use_madv_dontdump=1
+                  systemd.setenv=UBSAN_OPTIONS=print_stacktrace=1:print_summary=1:halt_on_error=1
```

Comment:
I don't think `systemd.setenv` affects PID1 itself. To pass those variables to `systemd` they should be exported before running `/sbin/init`. Another option would be to set them using `ManagerEnvironment` but last time I checked it wasn't exactly reliable in the sense that variables set via that setting got reset somewhere and appeared on `daemon-reexec` only.

To make it easier to figure out why PID1 and journald fail under ASan/UBsan `log_path` should probably be set as well. It kind of works even without that as long as there are no issues there but it might come in handy when new bugs are introduced.

Reply:
Hmm, for interactive use I'd expect the journal to suffice really. It's great to have all the information in one place when debugging something

Reply:
> I'd expect the journal to suffice really

`systemd-journald` redirects its ASan/UBSan reports straight to `/dev/null` by default so if they aren't captured with `log_path` they end up there with no way to view them. `systemd` can send its ASan/UBSan reports to different places depending on where exactly it fails. They can end up in `/dev/kmsg` (if it isn't run in containers), they can be sent to the console bypassing `journald` altogether or to `/dev/null`.

> It's great to have all the information in one place when debugging something

It is but in practice it doesn't always work when `journald` itself fails or `systemd` crashes before `journald` is started or after it's stopped.

Reply:
Granted, there are failure modes where `log_path` isn't helpful either but at least with it most backtraces are captured.

Reply:
> systemd-journald redirects its ASan/UBSan reports straight to /dev/null by default so if they aren't captured with log_path they end up there with no way to view them. systemd can send its ASan/UBSan reports to different places depending on where exactly it fails. They can end up in /dev/kmsg (if it isn't run in containers), they can be sent to the console bypassing journald altogether or to /dev/null.

This also depends on where systemd-journald is run. If it has access to /dev/kmsg, it will log there as well (and those messages) eventually end up in the journal.

I get the point that the journal might not always be accessible though. Ideally we could have asan log to both stdout and a log file. But log_path doesn't take multiple values. There's a log_to_syslog option as well, though if used for journald it might introduce a log cycle easily in journald

Reply:
> This also depends on where systemd-journald is run. If it has access to /dev/kmsg, it will log there as well (and those messages) eventually end up in the journal.

ASan/UBsan reports aren't logged with `log_*` (used by `systemd`). They are logged to `stderr` by `ASan/UBsan` and if `stderr` points to `/dev/null` (which is set by default in `systemd-journald.service`) it is supposed to always go to `/dev/null`. If those reports somehow end up in `/dev/kmsg` it would be a bug in systemd because it would mean that it failed to set up `StandardOutput/StandardError` correctly.

> But log_path doesn't take multiple values

It doesn't but pids are added at the end.

Reply:
> ASan/UBsan reports aren't logged with log_* (used by systemd). They are logged to stderr by ASan/UBsan and if stderr points to /dev/null (which is set by default in systemd-journald.service) it is supposed to always go to /dev/null. If those reports somehow end up in /dev/kmsg it would be a bug in systemd because it would mean that it failed to set up StandardOutput/StandardError correctly.

You're right, I wrongly assumed everything in journald would go to kmsg

I would do it like this, when we're in CI mode, we set log_path, if we're in regular interactive mode, we have ASAN log to stdout, is that a good compromise? We already check the CI_BUILD environment variable in the build script so there's a precedence for this

Reply:
> when we're in CI mode, we set log_path, if we're in regular interactive mode, we have ASAN log to stdout, is that a good compromise?

`StandardOutput` is set to `null` in `systemd-journald.service` so ASan/UBsan reports would be lost as well. It could work if `StandardOutput` was sent to a file (and the test suite does that too in addition to setting `log_path` to get around a bug in `libasan`) but in a way that would be the same as always setting `log_path`.

(it could be that I misunderstood what "we have ASAN log to stdout" means)

Reply:
Sounds good.

I'm not sure how to solve the issue in interactive mode for journald though. Ideally we could tell asan to log to kmsg just like the rest of journald. We could connect stdout to a file but that requires the developer to explicitly look into that file to get the leaks.

Another idea is to connect stdout and stderr directly to /dev/kmsg, journal should automatically pick all those messages up and store them in the journal

Reply:
I think the problem is that when `journald` fails under ASan/UBSan it tends to fail hard and get rate-limited by systemd eventually so it never picks up `/dev/kmsg`. Given that in interactive mode there should always be console probably I think setting `StandardOutput` to `kmsg+console` should suffice in the sense that those backtraces would be hard to miss.

Reply:
The problem with that is that `kmsg+console` is implemented by forwarding logs to the journald socket so it wouldn't really work either if journald breaks down.

We could try `StandardOutput=file:/dev/console` though I guess.

Reply:
This seems to work even better:

```
[root@fedora ~]# systemctl kill -s TERM systemd-journald
Received SIGTERM from PID 1 (systemd).
[root@fedora ~]# 
=================================================================
==386==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 100 byte(s) in 1 object(s) allocated from:
    #0 0x7fed008c268f in __interceptor_malloc (/lib64/libasan.so.8+0xba68f)
    #1 0x4089cf in main ../src/src/journal/journald.c:31
    #2 0x7fecfe0af54f in __libc_start_call_main (/lib64/libc.so.6+0x2954f)

SUMMARY: AddressSanitizer: 100 byte(s) leaked in 1 allocation(s).
```

Reply:
In systemd-nspawn the logs only appear on the console after a systemd-journald restart though.

```
[root@image ~]# systemctl kill -s TERM systemd-journald
[root@image ~]# systemd-journald.service: Scheduled restart job, restart counter is at 1.

[root@image ~]# systemctl kill -s TERM systemd-journald
Received SIGTERM from PID 1 (systemd).
[root@image ~]# 
=================================================================
==134==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 100 byte(s) in 1 object(s) allocated from:
    #0 0x7f11a0a3368f in __interceptor_malloc (/lib64/libasan.so.8+0xba68f)
    #1 0x4089cf in main ../src/src/journal/journald.c:31
    #2 0x7f119e22054f in __libc_start_call_main (/lib64/libc.so.6+0x2954f)

SUMMARY: AddressSanitizer: 100 byte(s) leaked in 1 allocation(s).
systemd-journald.service: Scheduled restart job, restart counter is at 2.
```

I switched to `StandardOutput=tty` which defaults to /dev/console

Reply:
I added the dropin to the PR

Reply:
I fixed the nspawn issue as well. Turns out both systemd and login were calling vhangup() on /dev/console, making ASAN unable to write the logs to /dev/console. Stopping both of those from calling vhangup() fixed the issue.

Reply:
Thanks! I've just installed `mkosi` and built `systemd` but I got distracted by a bunch of skipped tests. Looks like some dependencies aren't installed there. It has nothing to do with this PR though so I'll get back to that later once all strategically injected integer overflows pop up.